### PR TITLE
Add `media-query-no-invalid`

### PR DESCRIPTION
**Pull request by romainmenke**
> Which issue, if any, is this issue related to?

Closes https://github.com/stylelint/stylelint/issues/6916

> Is there anything in the PR that needs further explanation?

I've added some comments where I thought further explanation might be needed.

### CODE REVIEW

File: lib/rules/media-feature-name-value-no-unknown/index.js [hunk 1]
```diff
@@ -239,6 +240,8 @@ const rule = (primary) => {
 			};
 
 			parseMediaQuery(atRule).forEach((mediaQuery) => {
+				if (isMediaQueryInvalid(mediaQuery)) return;
```
Comment (romainmenke) [inline lib/rules/media-feature-name-value-no-unknown/index.js:243]: This rule should never warn for invalid media queries.
It should only check feature name and value pairs within grammatically valid media queries.

Otherwise it would overlap with the new rule.

File: lib/rules/media-feature-name-no-unknown/__tests__/index.js [hunk 2]
```diff
@@ -62,6 +62,9 @@ testRule({
 		{
 			code: '@media () { }',
 		},
+		{
+			code: '@media (grid: 1) {}',
+		},
```
Comment (romainmenke) [inline lib/rules/media-feature-name-no-unknown/__tests__/index.js:67]: Extra test coverage to check the boundaries of this rule vs. the new rule.

File: lib/rules/media-feature-no-invalid/README.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+Disallow invalid media queries.
+
+<!-- prettier-ignore -->
+```css
+@media not(min-width: 300px) {}
+/**    ↑
+ * This media query */
+```
+
+Media queries must be grammatically valid according to the [Media Queries Level 5](https://www.w3.org/TR/mediaqueries-5/) specification.
```
Comment (romainmenke) [inline lib/rules/media-feature-no-invalid/README.md:12]: I am unsure if this is the best way to explain this rule.  
I know this feature too well on a deeply technical level, so it isn't easy for me to explain it well for CSS authors.

Suggestions to improve would be very welcome :)

File: package.json [hunk 4]
```diff
@@ -130,7 +130,7 @@
   "dependencies": {
     "@csstools/css-parser-algorithms": "^2.2.0",
     "@csstools/css-tokenizer": "^2.1.1",
-    "@csstools/media-query-list-parser": "^2.1.0",
+    "@csstools/media-query-list-parser": "^2.1.1",
```
Comment (romainmenke) [inline package.json:133]: see : https://github.com/csstools/postcss-plugins/blob/main/packages/media-query-list-parser/CHANGELOG.md#211

> Fix parsing of `(width == 100px)`. This was erroneously parsed as a range query and will now instead be a general enclosed node.

File: lib/rules/media-query-no-invalid/README.md [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+Disallow invalid media queries.
+
+<!-- prettier-ignore -->
+```css
+@media not(min-width: 300px) {}
+/**    ↑
+ * This media query */
+```
+
+Media queries must be grammatically valid according to the [Media Queries Level 5](https://www.w3.org/TR/mediaqueries-5/) specification.
```
Comment (romainmenke) [inline lib/rules/media-query-no-invalid/README.md:12]: I am unsure if this is the best way to explain this rule.
I know this feature too well on a deeply technical level, so it isn't easy for me to explain it well for CSS authors.

Suggestions to improve would be very welcome :)

File: docs/user-guide/rules.md [hunk 6]
```diff
@@ -62,6 +62,7 @@ Disallow invalid syntax with these (sometimes implicit) `no-invalid` rules.
 | [`no-invalid-double-slash-comments`](../../lib/rules/no-invalid-double-slash-comments/README.md)<br/>Disallow invalid double-slash comments. | ✅ | |
 | [`no-invalid-position-at-import-rule`](../../lib/rules/no-invalid-position-at-import-rule/README.md)<br/>Disallow invalid position `@import` rules. | ✅ | |
 | [`string-no-newline`](../../lib/rules/string-no-newline/README.md)<br/>Disallow invalid newlines within strings. | ✅ | |
+| [`media-query-no-invalid`](../../lib/rules/media-query-no-invalid/README.md)<br/>Disallow invalid media queries. | | |
```
Comment (ybiquitous) [inline docs/user-guide/rules.md:65]: [suggest] Can you put it alphabetically, please?

File: lib/rules/index.js [hunk 7]
```diff
@@ -203,6 +203,7 @@ const rules = {
 	'media-feature-name-value-no-unknown': importLazy(() =>
 		require('./media-feature-name-value-no-unknown'),
 	)(),
+	'media-query-no-invalid': importLazy(() => require('./media-query-no-invalid'))(),
```
Comment (ybiquitous) [inline lib/rules/index.js:206]: [suggest] Please put it alphabetically also here?

File: lib/rules/media-query-no-invalid/index.js [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+const ruleName = 'media-query-no-invalid';
+
+const messages = ruleMessages(ruleName, {
+	rejected: (query) => `Unexpected invalid media query "${query}"`,
+});
+
+const HAS_MIN_MAX_PREFIX = /^(?:min|max)-/i;
+
+const meta = {
+	url: 'https://stylelint.io/user-guide/rules/media-feature-no-invalid',
```
Comment (ybiquitous) [inline lib/rules/media-query-no-invalid/index.js:28]: [suggest] Please fix the typo?

```suggestion
	url: 'https://stylelint.io/user-guide/rules/media-query-no-invalid',
```

Note: I've opened #6965 to check such a typo.

**Review by ybiquitous — CHANGES_REQUESTED** [review]
@romainmenke Thanks for the pull request. This is fantastic work!

I've commented a few on trivial things, but I believe there're no big problems 👍🏼

File: lib/rules/media-query-no-invalid/README.md (hunk 5 above)
Comment (Mouvedia) [reply lib/rules/media-query-no-invalid/README.md:12]: If I understand correctly this rule checks everything between `@media` and `{`, correct?

File: lib/rules/media-query-no-invalid/index.js (hunk 8 above)
Comment (romainmenke) [reply lib/rules/media-query-no-invalid/index.js]: Thank you!

File: lib/rules/media-query-no-invalid/README.md (hunk 5 above)
Comment (romainmenke) [reply lib/rules/media-query-no-invalid/README.md:12]: Yes, but at a different level, or from a different perspective than other rules.

This rule doesn't check :
- feature names (`width`, `hover`, ...)
- feature values (`10px`, `more`, `dark`, ...)
- feature name and value pairs (`width: 10px`)

You could say it only checks the structure.

Comment (Mouvedia) [reply lib/rules/media-query-no-invalid/README.md:12]: Alright then, if I had to suggest something Id say add a reminder that list the rules that check those.
As a user Id expect this rule to call them internally.

Comment (ybiquitous) [reply lib/rules/media-query-no-invalid/README.md:12]: Surely, it might be helpful to provide such a reminder, e.g., `media-feature-name-no-unknown`, `media-feature-name-value-no-unknown`, etc.

Comment (romainmenke) [reply lib/rules/media-query-no-invalid/README.md:12]: Good idea, listing the rules that cover other parts makes it clearer.

I've updated the rule documentation.

File: lib/rules/media-query-no-invalid/index.js [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+				if (isMediaQueryInvalid(mediaQuery)) {
+					// Queries that fail to parse are invalid.
+					invalidNodes.push(mediaQuery);
+
+					return;
+				}
+
+				mediaQuery.walk((entry) => {
+					// All general enclosured nodes are invalid.
+					if (isGeneralEnclosed(entry.node)) {
```
Comment (Mouvedia) [inline lib/rules/media-query-no-invalid/index.js:54]: minor

add `const node = entry.node;`
it's being accessed 8 times

File: lib/rules/media-feature-no-invalid/README.md (hunk 3 above)
Comment (Mouvedia) [reply lib/rules/media-feature-no-invalid/README.md]: If you think that this rule will be updated immediately after the release of the hypothetical level 6 specification then you can leave it as is.
If it's unlikely then something more vague would give us more leeway if that happened.

**Review by ybiquitous — APPROVED** [review]
Thanks, looks good 👍🏼

File: lib/rules/media-query-no-invalid/index.js (hunk 9 above)
Comment (romainmenke) [reply lib/rules/media-query-no-invalid/index.js:54]: Thank you, I will keep this in mind!
